x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/InferenceMetadataFieldMapper.java

```diff
@@ -0,0 +1,427 @@
+        // parse eagerly to extract the model settings first
+        Map<String, Object> map = parser.mapOrdered();
+        Object modelSettingsObj = map.remove(SemanticTextModelSettings.NAME);
+        if (modelSettingsObj == null) {
+            throw new DocumentParsingException(
+                parser.getTokenLocation(),
+                Strings.format(
+                    "Missing required [%s] for field [%s] of type [%s]",
+                    SemanticTextModelSettings.NAME,
+                    fieldName,
+                    SemanticTextFieldMapper.CONTENT_TYPE
+                )
+            );
+        }
+        Map<String, Object> modelSettingsMap = XContentMapValues.nodeMapValue(modelSettingsObj, "model_settings");
+        final SemanticTextModelSettings modelSettings;
+        try {
+            modelSettings = SemanticTextModelSettings.parse(
+                XContentHelper.mapToXContentParser(XContentParserConfiguration.EMPTY, modelSettingsMap)
+            );
+        } catch (Exception exc) {
+            throw new DocumentParsingException(
+                xContentLocation,
+                Strings.format(
+                    "Error parsing [%s] for field [%s] of type [%s]",
+                    SemanticTextModelSettings.NAME,
+                    fieldName,
+                    SemanticTextFieldMapper.CONTENT_TYPE
+                ),
+                exc
+            );
+        }
+        var nestedObjectMapper = updateSemanticTextFieldMapper(
+            context,
+            mapperBuilderContext,
+            res.parent,
+            (SemanticTextFieldMapper) res.mapper,
+            modelSettings,
+            xContentLocation
+        );
+
+        // we know the model settings, so we can (re) parse the results array now
+        XContentParser subParser = new MapXContentParser(
+            NamedXContentRegistry.EMPTY,
+            DeprecationHandler.IGNORE_DEPRECATIONS,
+            map,
+            XContentType.JSON
+        );
+        DocumentParserContext mapContext = context.switchParser(subParser);
+        parseFieldInferenceObject(xContentLocation, subParser, mapContext, nestedObjectMapper);
+    }
+
+    private void parseFieldInferenceObject(
+        XContentLocation xContentLocation,
+        XContentParser parser,
+        DocumentParserContext context,
+        NestedObjectMapper nestedMapper
+    ) throws IOException {
+        parser.nextToken();
+        failIfTokenIsNot(xContentLocation, parser, XContentParser.Token.START_OBJECT);
+        for (XContentParser.Token token = parser.nextToken(); token != XContentParser.Token.END_OBJECT; token = parser.nextToken()) {
+            switch (parser.currentName()) {
+                case RESULTS -> parseResultsList(xContentLocation, parser, context, nestedMapper);
+                default -> throw new DocumentParsingException(xContentLocation, "Unknown field name " + parser.currentName());
```

Comment:
This PR preserves this behaviour but only for the `_inference.$field_name.results` object. I am not a fan of this leniency but it is needed since we'll want to extend the information we provide in the future. I thought that we can be strict on the top level fields here since they should not change.